fearless/Common/Services/StorageQueryService/StorageKeysQueryService.swift

```diff
@@ -0,0 +1,169 @@
+    let mapper: AnyMapper<Data, T>
+
+    private var state: State = .none
+    private var completionClosure: ((Result<ResultType, Error>) -> Void)?
+    private weak var currentOperation: Operation?
+    private var mutex = NSLock()
+
+    init(
+        connection: JSONRPCEngine,
+        operationManager: OperationManagerProtocol,
+        prefixKeyClosure: @escaping () throws -> Data,
+        mapper: AnyMapper<Data, T>,
+        pageSize: UInt32 = 1000
+    ) {
+        self.connection = connection
+        self.operationManager = operationManager
+        self.prefixKeyClosure = prefixKeyClosure
+        self.pageSize = pageSize
+        self.mapper = mapper
+    }
+
+    private func loadNext() {
+        guard case let .inProgress(_, lastKey) = state else {
+            return
+        }
+
+        do {
+            let prefixKey = try prefixKeyClosure()
+
+            let request = PagedKeysRequest(
+                key: prefixKey.toHex(includePrefix: true),
+                count: pageSize,
+                offset: lastKey.map { $0 }
+            )
+
+            let operation = JSONRPCOperation<PagedKeysRequest, [String]>(
+                engine: connection,
+                method: RPCMethod.getStorageKeysPaged,
+                parameters: request
+            )
+
+            currentOperation = operation
+
+            operation.completionBlock = { [weak self] in
+                DispatchQueue.global().async {
+                    do {
+                        let response = try operation.extractNoCancellableResultData()
+                        self?.handlePage(response: response)
+                    } catch {
+                        self?.completeWithError(error)
+                    }
+                }
+            }
+
+            operationManager.enqueue(operations: [operation], in: .transient)
+
+        } catch {
+            completeWithError(error)
+        }
+    }
+
+    private func handlePage(response: [String]) {
+        mutex.lock()
+
+        defer {
+            mutex.unlock()
+        }
+
+        guard case let .inProgress(oldItems, _) = state else {
+            return
+        }
+
+        do {
+            let mappedItems: [T] = try response.map { hexKey in
+                let keyData = try Data(hexString: hexKey)
+                return mapper.map(input: keyData)
+            }
+
+            let resultItems = oldItems + mappedItems
+
+            if mappedItems.count < pageSize {
+                completeWithResult(resultItems)
+            } else {
+                state = .inProgress(partialResult: resultItems, lastKey: response.last)
+                loadNext()
+            }
+        } catch {
+            completeWithError(error)
+        }
+    }
+
+    private func completeWithResult(_ result: [T]) {
+        let closure = completionClosure
+
+        completionClosure = nil
+        state = .completed(result: result)
+
+        closure?(.success(result))
```

Comment:
```swift
state = .completed(result: result)       
completionClosure?(.success(result))
completionClosure = nil

Reply:
The main purpose here is that we don't have any assumption that the service will not be called in closure